File: dynamic/140_wordbreak/Solution1.java
import java.io.*;
import java.util.*;
import java.text.*;
import java.math.*;
import java.util.regex.*;

// class TreeNode {
// 	int val;
// 	TreeNode left;
// 	TreeNode right;
// 	TreeNode(int x) { val = x; }
// }
// private void createTree() {
// 	//       5
// 	//    2      7
// 	//  1   3      10
// 	TreeNode node5 = new TreeNode(5);
// 	TreeNode node2 = new TreeNode(2);
// 	TreeNode node1 = new TreeNode(1);
// 	TreeNode node7 = new TreeNode(7);
// 	TreeNode node10 = new TreeNode(10);

// 	node5.left = node2;
// 	node5.right = node7;
// 	node2.left = node1;
// 	node2.right = node3;
// 	node5.right = node7;
// 	node7.right = node10; 
// 	return node5;
// }
// private void printList(List<String> list) {
// 	for(String s: list) {
// 		System.out.println(s);
// 	}
// }
// private void printQueue(Queue<String> list) {
// 	for(String s: list) {
// 		System.out.println(s);
// 	}
// } 
// private void printListOfList(List<List<Integer>> listOflist) {
// 	for(List<Integer> list: listOflist){
// 		for(int n : list) {
// 			System.out.print(n + ", ");
// 		}
// 		System.out.println();
// 	}
// }
// private void printMatrix(int[][] matrix) {
//     for(int[] s: matrix) {
//         for(int n: s) {
//             System.out.print(n + ",'");
//         }
//     }
//     System.out.println();
// }  

public class Solution1 {

    public List<String> wordBreak(String s, List<String> wordDict) {
        int[][] dp = new int[s.length()][s.length() + 1];
        for (int i = 0; i < dp.length; i++) {
            for(int j = 0; j < dp[i].length; j++) {
                dp[i][j] = 0;
            }
        }

        for (int i = 1; i <= s.length(); i++) {
            for(int j = 0; j < s.length() - i + 1 ; j++) {
                // System.out.println("j:" + j + ", i:" + i);
                if(listContainsWord(wordDict, s.substring(j, j + i)) == true) {
                    dp[j][i] = 1;
                    continue;
                }

                for (int k = 1; k < i && k < s.length() - j; k++) {
                    if(dp[j][k] != 0 && dp[j + k][i - k] != 0) {
                        dp[j][i] = 2;
                        break;
                    }
                }
            }
        }


        return null;
    }

    private void printMatrix(int[][] matrix) {
        for(int[] s: matrix) {
            for(int n: s) {
                System.out.print(n + ",");
            }
            System.out.println();            
        }

    }    

    private boolean listContainsWord(List<String> wordDict, String s) {
    	System.out.println(s);
    	for (String word : wordDict) {
    		if (s.equals(word)) {
    			return true;
    		}
    	}
    	return false;
    }

	public static void main(String[] args) {
		Solution1 obj = new Solution1();	
		List<String> list = new ArrayList<String>();
		list.add("cat");
		list.add("cats");
        list.add("and");
        list.add("sand");
        list.add("dog");
        obj.wordBreak("catsanddog", list);
	}

}

File: dynamic/140_wordbreak/Solution1.cpp
#include <algorithm>
#include <iostream>
#include <vector>
#include <string>
using namespace std;

class Solution {
public:
    
    vector<vector<int> >dp;
    vector<string> vals;
    string val;

    vector<string> wordBreak(string s, vector<string>& wordDict) {
        int len = (int)s.size();
        dp.resize(len);
        for(int i = 0; i < len; i++) {
            dp[i].resize(len + 1, 0);
        }

        for(int i = 1; i <= len; i++) {
            for(int j = 0; j < len -i + 1; j++) {
                //直接存在于字典中，是第1种情况
                
                if(std::find(wordDict.begin(), wordDict.end(), s.substr(j, i)) != wordDict.end()) {
                    dp[j][i] = 1;
                    continue;
                }
                //如果不存在，则看子串是不是能被切分，这是第2中情况
                for(int k = 1; k < i && k < len -j; k++) {
                    if(dp[j][k] && dp[j + k][i - k]) {
                        dp[j][i] = 2;
                        break;
                    }
                }
            }
        }

        //不能切分，不用dfs了
        if(dp[0][len] == 0) {
            return vals;
        }
        printVectorOfVector(dp);
        dfs(s, 0);
        return vals;
    }

    void printVectorOfVector(vector< vector<int> >& v) {

        for(std::vector< vector<int> >::iterator it = v.begin(); it != v.end(); ++it) {
           for(std::vector<int>::iterator itOfit = (*it).begin(); itOfit != (*it).end(); ++itOfit) {
                std::cout << ' ' << *itOfit;
            } 
            std::cout<< endl;
        } 
    }

    void dfs(const string& s, int start) {
        int len = (int)s.size();
        if(start == len) {
            vals.push_back(val);
            return;
        }

        for(int i = 1; i <= len - start;i++) {
            if(dp[start][i] == 1) {
                int oldLen = (int)val.size();
                if(oldLen != 0) {
                    val.append(" ");
                }
                val.append(s.substr(start, i));

                //我们从start + i开始继续dfs
                dfs(s, start + i);
                val.erase(oldLen, string::npos);
            }
        }
    } 


};

int main(){
    Solution solution;
    std:string s1 ("catsanddog");
    vector<string> list;
    list.push_back("cat");
    list.push_back("cats");
    list.push_back("and");
    list.push_back("sand");
    list.push_back("dog");
    solution.wordBreak(s1, list);
}